src/source/src/dml_manager.rs

```diff
@@ -32,35 +34,63 @@ pub type DmlManagerRef = Arc<DmlManager>;
 /// channel instead of offering a `write_chunk` interface).
 #[derive(Default, Debug)]
 pub struct DmlManager {
-    batch_dmls: Mutex<HashMap<TableId, TableSourceRef>>,
+    batch_dmls: RwLock<HashMap<TableId, Weak<TableSource>>>,
 }
 
 impl DmlManager {
     pub fn new() -> Self {
         Self {
-            batch_dmls: Mutex::new(HashMap::new()),
+            batch_dmls: RwLock::new(HashMap::new()),
         }
     }
 
     pub fn register_reader(
         &self,
-        table_id: &TableId,
+        table_id: TableId,
         column_descs: &[ColumnDesc],
-    ) -> TableSourceRef {
-        let mut batch_dmls = self.batch_dmls.lock();
-        if !batch_dmls.contains_key(table_id) {
-            let batch_dml = Arc::new(TableSource::new(column_descs.to_vec()));
-            batch_dmls.insert(*table_id, batch_dml);
+    ) -> Result<TableSourceRef> {
+        let mut batch_dmls = self.batch_dmls.write();
+        match batch_dmls.entry(table_id) {
+            Entry::Occupied(o) => o.get().upgrade().ok_or_else(|| {
```

Comment:
Is there any way to clean up the invalid weak ref entries?

Reply:
https://github.com/risingwavelabs/risingwave/blob/820ff8c95be38469d5757751962f92046d299f98/src/source/src/manager.rs#L138 for example